### PR TITLE
Optimized Inventory, Cart Inventory, and Storage usage

### LINKED ISSUE #441: Cart Items need be saved before vending
11251288_820368931372402_74167565396148733_n 1

### PR DESCRIPTION
**Pull request by aleos89**
- These storage types now mimic Guild Storage and can be loaded/saved whenever needed.
- Relieves mmo_charstatus from having to send storage types and lets the char-server handle it.
- All storage types now have an increased max amount.
- Fixes #441 - Players will no longer be required to log out to resync cart item data before opening a Vending Store.
- Refactored player weight and cart weight calculations into their own functions.
- Added script commands guildstoragecountitem[2] and guildstoragedelitem[2].
- Refactored several function return types as well as documentation.

Thanks to @lighta and @cydh for their help with it!

### CODE REVIEW

**Comment by cydh** [comment]
new? any difference with #1101?

**Comment by Lemongrass3110** [comment]
@cydh as @aleos89 stated on the old PR he cleaned it up. The other pull request ended up having two parents which most likely caused that weird behaviour of github showing some commits and some of the comments on the commits on that pull request.
So basically everything is just the same, but now it is readable and comparable.
If you want us to take something into account please comment it here now.
Thank you for understanding.

**Comment by stealth62** [comment]
i notice in intif.c
```
static void intif_parse_StorageSaved(int fd)
```
case TABLE_STORAGE: //storage
			if (sd && sd->state.storage_flag == 1)
				sd->state.storage_flag = 0; //  <<<<< this cause storage hang and cant close. if storage is open for a few minutes
			//ShowInfo("Storage has been saved (AID: %d).\n", RFIFOL(fd, 2));
			break;
its safe to remove any side effect? in my testing i remove the line and my storage back to normal not causing hang i can properly close the storage.

**Comment by stealth62** [comment]
before
```
case TABLE_STORAGE: //storage
            if (sd && sd->state.storage_flag == 1)
                sd->state.storage_flag = 0; //  <<<<< if storage is open for a few minutes the  storage hang and can't close. 
            //ShowInfo("Storage has been saved (AID: %d).\n", RFIFOL(fd, 2));
            break;
```
after
```
case TABLE_STORAGE: //storage
           if (sd->state.storage_flag == 1)
       intif_storage_save(sd,TABLE_STORAGE);
            //ShowInfo("Storage has been saved (AID: %d).\n", RFIFOL(fd, 2));
            break;
```
is this correct sir?

**Comment by aleos89** [comment]
No, the block of code I commented on was from `chrif_save`. So like what you said before, just commenting out the line should restore it back to normal. Every time the character is saved (which is where I didn't wait long enough while testing) it will call `intif_parse_StorageSaved()` and then see that the flag is 1 because the storage is still open and will set it to 0. This is what causes you to get stuck. Since in `storage_storage_quit()` we remove the flag at the end as well, the check in `intif_parse_StorageSaved` is not really needed.

**Comment by stealth62** [comment]
ok. so its safe to remove that line. thank you very much sir @aleos89

**Comment by stealth62** [comment]
Hi sir @aleos89 i found another problem if you have item in cart for example 1pc apple weight 2. In cart information qty 1/100 and weight 2/10500 then my character relogs the cart information says qty 0/100 and weight 0/10500 then i tried to put in my inventory the cart weight calc becomes qty -1/100  weight -2/10500. I found out in status.c
int status_calc_pc_(struct map_session_data* sd, enum e_status_calc_opt opt)
```
----- MISC CALCULATIONS -----
	//weight
	status_calc_weight(sd,2);
	status_calc_cart_weight(sd,2);
```
i change to
----- MISC CALCULATIONS -----
	//weight
	status_calc_weight(sd,1|2);
	status_calc_cart_weight(sd,1|2);
```
and it works the item weight in cart information updated even you relogs.

**Comment by stealth62** [comment]
Hi sir @aleos89 i found another problem if you have item in cart for example 1pc apple weight 2. In cart information qty 1/100 and weight 2/10500 then character relogs the cart information says qty 0/100 and weight 0/10500 then i tried to put the apple in my inventory the cart weight calc becomes qty -1/100  weight -2/10500 which is bug. I found out in status.c
int status_calc_pc_(struct map_session_data* sd, enum e_status_calc_opt opt)
```
if (opt&SCO_FIRST) {
		// Load Hp/SP from char-received data.
		sd->battle_status.hp = sd->status.hp;
		sd->battle_status.sp = sd->status.sp;
		sd->regen.sregen = &sd->sregen;
		sd->regen.ssregen = &sd->ssregen;
status_calc_weight(sd, 1);
	status_calc_cart_weight(sd, 1);  <<< somethings wrong when code putted here opt&SCO_FIRST
	}
```
change to
```
status_calc_weight(sd, 1);
	status_calc_cart_weight(sd, 1); <<<<<<< this will update cart calculation info even in relogs.
if (opt&SCO_FIRST) {
		// Load Hp/SP from char-received data.
		sd->battle_status.hp = sd->status.hp;
		sd->battle_status.sp = sd->status.sp;
		sd->regen.sregen = &sd->sregen;
		sd->regen.ssregen = &sd->ssregen;
	}
and
----- MISC CALCULATIONS -----
	//weight
	status_calc_weight(sd,2);
	status_calc_cart_weight(sd,2);
```
change to
```
----- MISC CALCULATIONS -----
	//weight
	status_calc_weight(sd,2);
	status_calc_cart_weight(sd,1|2);//item from cart to inventory vice versa will update cart info calculation correctly.
```

**Comment by CairoLee** [comment]
Hi sir @aleos89 i found new problem. 

**Step 1:** Create an new Merchant and keep the character a cart (keep the cart is empty).
**Step 2:** ```@item 501\502\503\504``` and put them into the cart
**Step 3:** relog
**Step 4:** ```@item 5113``` and put it into the cart
**Step 5:** use ```@allskill``` and ```@go 0``` to prontera
**Step 6:** Open an vending and sell the 5113.
**Step 7:** After open the vending, lookup the vending list, you will see vending item is 501**(not 5113)**.

**Comment by CairoLee** [comment]
Hi @aleos89  any news? I'm looking forward to this .

**Comment by aleos89** [comment]
Sorry, I've been busy with personal things recently so during my free time I don't have the chance to load up RO and go through this to test further. I should be relieved by next week though to begin pushing through this PR again.

**Comment by aleos89** [comment]
@stealth62: I am unable to reproduce this issue. I swapped items around and relogged several times. My weight was always calculated properly.

@CairoLee: I am unable to reproduce this issue as well. I created several items, relogged, made more and opened a shop. The item showed up just fine to me and other players. I purchased the item and it gave me the correct item.

**Comment by RadianFord** [comment]
@aleos89 how can i apply this on the master?

**Comment by RadianFord** [comment]
I tried using this
Hash: 55165ac0bfce710399ffea9b0c62aefc4904cd3c
Pre Renewal
2013 12 23 client

Its working fine for me.

**Comment by CairoLee** [comment]
@aleos89 okay, sir.
I will try to reproduce it again. maybe i can record an video to show it. XD

**Comment by RadianFord** [comment]
@aleos89 sorry to bother you but when i compile my trunk on linux i got this warnings.
untitled

**Comment by aleos89** [comment]
I can fix that up. This PR has so conflicts with merging against master so I need to resolve that as well.

**Comment by RadianFord** [comment]
@aleos89 cool thanks!

**Comment by CairoLee** [comment]
@aleos89 Sorry i'm late.

I merge this request in master. and test it again. still have same problem. I record an video, and publish the video to youtube. Hope it can help.

https://www.youtube.com/watch?v=1D-8aThxhZY

#### Reproduce

1.New character
2.@job 5
3.@cart 1
4.@lvup 90 (make sure have sp..)
5.@go 0
6.relog
7.@item 501\502\503\504
8.put them into cart
9.relog again
10.@item 5113
11.put it into cart
12.@allskill
13.open a shop and sell 5113
14.in My Shop windows, you can see it's an Orange Potion...

**Comment by RadianFord** [comment]
@CairoLee Hi I did what you did in the video and i cant produced it,

**Comment by CairoLee** [comment]
@RadianFord Thank you. XD

Maybe it's associated with autosave..? Can you do it faster? Don't let the map-server do autosave... 

Here is my config when i record the video (in conf/map_athena.conf)... 
```
autosave_time: 300
minsave_time: 100
save_settings: 255
```

**Comment by RadianFord** [comment]
I am using this settings @CairoLee 
```
// Database autosave time
// All characters are saved on this time in seconds (example:
// autosave of 60 secs with 60 characters online -> one char is saved every 
// second)
autosave_time: 300

// Min database save intervals (in ms)
// Prevent saving characters faster than at this rate (prevents char-server 
// save-load getting too high as character-count increases)
minsave_time: 100

// Apart from the autosave_time, players will also get saved when involved
// in the following (add as needed):
// 1: after every successful trade
// 2: after opening vending/every vending transaction
// 4: after closing storage/guild storage.
// 8: After hatching/returning to egg a pet.
// 16: After successfully sending a mail with attachment
// 32: After successfully submitting an item for auction
// 64: After successfully get/delete/complete a quest
// 128: After every bank transaction (deposit/withdraw)
// NOTE: These settings decrease the chance of dupes/lost items when there's a
// server crash at the expense of increasing the map/char server lag. If your 
// server rarely crashes, but experiences interserver lag, you may want to set
// these off.
save_settings: 255
```

**Comment by Jeybla** [comment]
Merged against current master here: https://github.com/Jeybla/rathena/tree/cleanup/storages_optimization

I'd like to do some testing, so here is a testing TODO:
- [x] Put item in storage/cart
- [ ] Put many items in storage/cart
- [ ] Put more than max items in storage/cart
- [ ] Let storage opened for ~10 minutes; try to close
- [ ] Let storage opened for ~10 minutes; try to put/get items in/from it
- [ ] Put items from storage in cart; put from cart to storage **(works, but resets after logout?)**
- [ ] Vending Test
- [ ] Bulk Buyers Shop
- [ ] Drop/pick items
- [ ] Sell items
- [ ] Will add more soon

Everything should be tested twice. Once directly logging out and the other test should be by staying online for at least 5 Minutes after the action.
If you have anything else to test I'd really appreciate if you add more testing cases :)

**Comment by anacondaqq** [comment]
> Test bound items (charbound, accbound, guildbound)

This test should be most careful. Because there is guild bounded items, which returned back as i remember to guild leader storage. And rental items too.

File: src/common/mmo.h [hunk 1]
```diff
@@ -283,19 +283,26 @@ struct skill_cooldown_data {
...
-	bool locked; ///< If locked, can't use storage when item bound retrieval
-	uint32 opened; ///< Holds the char_id that open the storage
+struct s_storage {
+	bool dirty; ///< Dirty status, data needs to be saved
+	bool status; ///< Current status of storage (opened or closed)
+	int amount; ///< Amount of items in storage
+	bool lock; ///< If locked, can't use storage when item bound retrieval
+	uint32 id; ///< Account ID / Character ID / Guild ID (owner of storage)
+	enum storage_type type; ///< Type of storage (inventory, cart, storage, guild storage)
+	union { // Max for inventory, storage, cart, and guild storage are 1637 each without changing this struct and struct item [2014/10/27]
```
Comment (Jeybla) [inline src/common/mmo.h]: [2014/10/27]?

You did this two years ago? xD

File: src/map/intif.c [hunk 2]
```diff
@@ -3080,21 +3099,21 @@ void intif_itembound_guild_retrieve(uint32 char_id,uint32 account_id,int guild_i
  */
 void intif_parse_itembound_ack(int fd) {
 	int guild_id = RFIFOW(fd,6);
-	struct guild_storage *gstor = gstorage_get_storage(guild_id);
+	struct s_storage *gstor = guild2storage(guild_id);
```
Comment (Jeybla) [inline src/map/intif.c]: Small typo? Should be guild2storage imo :)

**Comment by Jeybla** [comment]
Okay I'm done with testing. I think it should be mostly save to push to the master with my little patches.
Issues I6-I9 aren't critical and can be fixed afterwards.

I tested the highest storage and guildstorage size. Unfortunately it's not much. The highest I've tested was 900. But however it works pretty good and it's better than the current 600 slots since random option release.
Maybe we can cleanup the mmo_charstatus struct somehow. 

Anyway thanks for the release @aleos89, @lighta and @cydh. Maybe I could help you a bit. Please have a look at https://github.com/Jeybla/rathena/tree/cleanup/storages_optimization. Thanks!

**Comment by cydh** [comment]
@Jeybla thank you for testing! /no1

16. I have no idea. But seems it's possible since when map-server is crashed, there's no way map-server asking for save to char/inter-server
17. Any report for this?
18. I think it's just custom message from our end, in official, there's no message if cannot move to cart/storage. But, they do something: Request to move -> Item removed from inventory -> If cannot be moved, item is added back to inventory. so the failed item to move will be in the last entry of inventory.
19. This is really need to be checked.

**Comment by aleos89** [comment]
Thanks for such a thorough test @Jeybla! I've began merging this to master and will look at the issues you brought up and the fixes. Thanks again!

**Comment by Jeybla** [comment]
@cydh 
1. Yes, I don't see a possible solution, too.
2. Sometimes the item price sold seems not to be correct.
3. Yep, just wanted to  suggest to change this message :)
4. Fix for I9:
```diff
index 1b4d8e3..2c16a02 100644
--- a/src/map/trade.c
+++ b/src/map/trade.c
@@ -385,7 +385,10 @@ void trade_tradeadditem(struct map_session_data *sd, short index, short amount)
                return;
        }
 
-       if( ((item->bound == BOUND_ACCOUNT || item->bound > BOUND_GUILD) || (item->bound == BOUND_GUILD && sd->status.guild_id != target_sd->status.guild_id)) && !pc_can_give_bounded_items(sd) ) { // Item Bound
+       if( item->bound > 0  && 
+               !(item->bound == BOUND_GUILD && sd->status.guild_id > 0 && sd->status.guild_id == target_sd->status.guild_id) && 
+               !(item->bound == BOUND_PARTY && sd->status.party_id > 0 && sd->status.party_id == target_sd->status.party_id) &&
+               !pc_can_give_bounded_items(sd) ) { // Item Bound
                clif_displaymessage(sd->fd, msg_txt(sd,293));
                clif_tradeitemok(sd, index+2, 1);
                return;
```

@aleos89 you're welcome! Maybe you want to compare it with my merge too. It's not easy to apply the merge correctly since many things changed since then. I did it twice to be sure m(
I tested everything on 4174ed7e50b8064210c3cf78a43fa8a68c81d10c. Good luck!

**Comment by anacondaqq** [comment]
About 

> #define MAX_STORAGE 600 ///Max number of storage slots a player can have

Can you please  tell, after applying this pull request, what limitations of MAX_STORAGE? What maximum value for this setting? Can i for example set here MAX_STORAGE 3000?

**Comment by aleos89** [comment]
I believe above Jeybla said he tested around 900 for storage and guild storage. I haven't tested in a while and will have to try it out myself again.

**Comment by Jeybla** [comment]
Yeah sadly it's not much more than before this pr and the random option items.

Maybe we should discuss about increasing the mmo_charstatus size later ^^

**Comment by cydh** [comment]
for vending issue, maybe better
  1. Check first if there's unsaved items
  2. Then
    * If there's no unsaved items, call `clif_openvendingreq`.
    * If there's unsaved items, ask for sync for `intif_storage_save(sd, TABLE_CART); // Save cart item data`. In the reply of `intif_storage_save`, do `intif_storage_request(sd, TABLE_CART); // Update cart item ID information`. But this is need extra execution time, it'll be faster for this case only, the `intif_storage_save` will returns the new item for cart.
  3. After cart data is gained, then do `clif_openvendingreq`

**Comment by cydh** [comment]
Super bump for #1551

**Comment by Jeybla** [comment]
Bump! Me and imo many others need this, to stay updated with rA since the random option system. Is there anything missing? Can I help you somehow?

Thx :>

**Comment by anacondaqq** [comment]
when i will finish some job first i will join to tests too, please don't hurry to merge such big changes to rAthena.

**Comment by cydh** [comment]
> please don't hurry to merge such big changes to rAthena.

we don't hurry for this, it's been 2 years since I init the changes, we can still wait for other year(s) xD

File: src/map/chrif.c [hunk 3]
```diff
@@ -298,9 +298,14 @@ int chrif_save(struct map_session_data *sd, int flag) {
 
 	chrif_bsdata_save(sd, (flag && (flag != 3)));
 
+	if (sd->state.storage_flag == 1)
+        intif_storage_save(sd,TABLE_STORAGE);
```
Comment (hnomkeng) [inline src/map/chrif.c]: warning C4013: 'intif_storage_save' undefined; assuming extern returning int

File: src/map/pc.c [hunk 4]
```diff
@@ -1468,12 +1456,9 @@ void pc_reg_received(struct map_session_data *sd)
 		clif_changeoption( &sd->bl );
 	}
 
-	pc_check_expiration(sd);
-
-	if( sd->state.autotrade ) {
-		clif_parse_LoadEndAck(sd->fd, sd);
-		sd->autotrade_tid = add_timer(gettick() + battle_config.feature_autotrade_open_delay, pc_autotrade_timer, sd->bl.id, 0);
-	}
+	intif_storage_request(sd,TABLE_STORAGE); // Request storage data
```
Comment (hnomkeng) [inline src/map/pc.c]: warning C4013: 'intif_storage_request' undefined; assuming extern returning int

File: src/map/skill.c [hunk 5]
```diff
@@ -7300,7 +7300,12 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			else {
 				sd->state.prevend = 1;
 				sd->state.workinprogress = WIP_DISABLE_ALL;
-				clif_openvendingreq(sd,2+skill_lv);
+				sd->vend_skill_lv = skill_lv;
+				ARR_FIND(0, MAX_CART, i, sd->cart.u.items_cart[i].nameid && sd->cart.u.items_cart[i].id == 0);
+				if (i < MAX_CART)
+					intif_storage_save(sd, TABLE_CART_);
```
Comment (hnomkeng) [inline src/map/skill.c]: warning C4013: 'intif_storage_save' undefined; assuming extern returning int

File: src/char/char.c [hunk 6]
```diff
@@ -601,12 +589,14 @@ int char_memitemdata_to_sql(const struct item items[], int max, int id, int tabl
 	SqlStmt_BindColumn(stmt, 7, SQLDT_UINT,      &item.expire_time, 0, NULL, NULL);
 	SqlStmt_BindColumn(stmt, 8, SQLDT_UINT,      &item.bound,       0, NULL, NULL);
 	SqlStmt_BindColumn(stmt, 9, SQLDT_UINT64,    &item.unique_id,   0, NULL, NULL);
+	if (tableswitch == TABLE_INVENTORY)
+		SqlStmt_BindColumn(stmt, 10, SQLDT_CHAR, &item.favorite,    0, NULL, NULL);
 	for( i = 0; i < MAX_SLOTS; ++i )
-		SqlStmt_BindColumn(stmt, 10+i, SQLDT_USHORT, &item.card[i], 0, NULL, NULL);
+		SqlStmt_BindColumn(stmt, 10+offset+i, SQLDT_USHORT, &item.card[i], 0, NULL, NULL);
```
Comment (aleos89) [inline src/char/char.c]: I think you merged incorrectly. Travis nor Appveyor show any compile warnings/errors.

File: src/map/chrif.c (hunk 3 above)
Comment (aleos89) [inline src/map/chrif.c]: Again, I think you merged incorrectly. Check your intif.h to make sure you defined the function in there.

File: src/map/pc.c (hunk 4 above)
Comment (aleos89) [inline src/map/pc.c]: Again, I think you merged incorrectly. Check your intif.h to make sure you defined the function in there.

File: src/map/skill.c (hunk 5 above)
Comment (aleos89) [inline src/map/skill.c]: Again, I think you merged incorrectly. Check your intif.h to make sure you defined the function in there.

**Comment by cydh** [comment]
> Follow up to 800714a forgot to eliminate TABLE_CART_ completely.

then, there's no point if do not revert 
```bool intif_storage_save(struct map_session_data *sd, enum storage_type type)```
to
```bool intif_storage_save(struct map_session_data *sd, struct s_storage *stor)```

**Pull request merged**